dawn/docs/parallel_model.rst

```diff
@@ -0,0 +1,251 @@
+
+Formal Definition of the Dawn Parallel Model
+============================================
+
+Contract to the user from HIR
+-----------------------------
+
+Dawn will generate code based on the operations passed from high-level
+intermediate representation. It's specification can be found `here`_.
+
+Dependencies
+~~~~~~~~~~~~
+
+Generally, stencil computation in dawn differentiates between horizontal
+and the vertical dimension. The assumption is that individual statements
+of a stencil are embarrassingly parallel in the horizontal. For that
+reason, there is no guarantee that any horizontal point's computation
+will be finished before any other in each user statement.
```

Comment:
> The assumption is that individual statements of a stencil are embarrassingly parallel in the horizontal.
This implies that you cannot do self-assignment in general (`out = out[i+1]`), but is there anything else?

Reply:
maybe we should rephrase it then. What it means is there can not be loop carried dependencies (reductions) in the horizontal.

Since we do field-versioning, we should probably explain it here?

Reply:
I see. Yes field versioning sounds like a good thing to mention here. So what should happen if I write
```
out = out[i+1] + out[i - 1]
```
If we do field versioning, this will become
```
out' = out[i+1] + out[i - 1]
```
Which looks perfectly legal. So the statement is not about a missing guarantee, but it is a statement that each statement is executed double-buffered
```
out' = out[i+1] + out[i-1]
out = out'
```

Reply:
I would propose something like
```
Generally, stencil computation in dawn differentiates between horizontal
and the vertical dimension. The assumption is that individual statements
of a stencil are embarrassingly parallel in the horizontal. For that
reason, there is no guarantee that any horizontal point's computation
will be finished before any other in each user statement.
Computations with data dependencies like 
tmp = field[i-1]
field = tmp[j+1] + tmp[j-1]
will be resolved by field versioning and adding synchronizations between statements with data dependencies. 
That means that iterative solvers where a computation accumulates values over one of the horizontal dimensions are not possible
```

Reply:
@lukasm91 
the current implementation does a different thing:
we currently have 
```
NewMS:
 out' = out
MS:
  out = out'[i+1] + out'[i-1]
```
so snyc is implied if versioning is required. 

@cosunae 
should we maybe touch on all the ways this can happen?
the`out = out[offset] + something` case is a "different one" that is also interesting, right?

Reply:
Agreed that the implementation might be different, but it doesn't matter (both interpretations are ok). The important thing is that it is double buffered.
```
out' = out[i+1] + out[i-1]
out = out'
```
and
```
out' = out
out = out'[i+1] + out'[i-1]
```
They should be equivalent (of course statement by statement, applied to the full ij-plane).

Reply:
true. And in the contract we have with the SIR it should not matter since the user only needs to know the effect (his field is versioned, his input is read properly and the output is in the correct field)